### ISSUE TITLE
FEATURE: Add KEYs to policies and executions

### ISSUE THREAD

**Issue by reisenberger** [App-vNext/Polly#139]
# Proposal: Add ids/keys to policies and executions

Ids or Keys could be useful in Polly for a number of functions, and at a number of levels of granularity.
## PolicyWrapKey

| Key type | PolicyWrapKey |
| --- | --- |
| Description / Purpose | A unique key for each Policy Wrap.  While PolicyKey varies with each policy in a wrap, PolicyWrapKey remains constant over the wrap. |
| Primary use | Retrieve a Policy Wrap from Registry? |
| Additional use | Include in emitted events/stats, to indicate policywrap in use |
| Default if not set | short-type-name+some-guid-part?; null if execution not part of a pipeline |
| Behaviour in PolicyWrap | Retains own value |
| How would it be specified? | extension method `.WithPolicyWrapKey(string policyWrapKey)` on `PolicyWrap`, but set-once-only enforced |
| How should it be exposed? | Defined as `PolicyWrap.PolicyWrapKey`?; exposed to execution as `Context.PolicyWrapKey` |
## PolicyKey

| Key type | Policy Key |
| --- | --- |
| Description / Purpose | A unique key for an individual configured Policy instance |
| Primary use | Retrieve a policy from registry |
| Additional use | Include in emitted events/stats, to indicate policy in use |
| Default if not set | short-type-name+some-guid-part ? |
| Behaviour in PolicyWrap | Retains own value |
| How would it be specified? | extension method `.WithPolicyKey(string policyKey)` on `Policy`, but set-once-only enforced |
| How should it be exposed? | Defined as `Policy.PolicyKey`? Exposed to execution as `Context.PolicyKey` |
## ExecutionKey

| Key type | ExecutionKey |
| --- | --- |
| Description / Purpose | A unique key for each _call site usage_ of a policy.  (A single policy instance may be re-used in multiple code locations (see [wiki](https://github.com/App-vNext/Polly/wiki/Some-policy-patterns))). |
| Primary use | Used by `CachePolicy` as the cache key for the given execution |
| Additional use | Include in emitted events/stats, to indicate _where_ policy in use |
| Default if not set | null? |
| Behaviour in PolicyWrap | Takes value passed by any preceding policy in the wrap |
| How would it be specified? | by passing a `new Context(executionKey)` with the Execute call |
| How should it be exposed? | `Context.ExecutionKey` |
## ExecutionGuid

| Key type | ExecutionGuid |
| --- | --- |
| Description / Purpose | A unique GUID for each _invocation_ through a policy |
| Primary use | Distinguish individual calls through the policy, so that stats-interpreting code can distinguish calls in multi-threaded environment |
| Additional use | . |
| Default if not set | Guid.NewGuid() |
| Behaviour in PolicyWrap | Takes value passed by any preceding policy in the wrap |
| How would it be specified? | not for the coder to specify - generated randomly by each call to Execute |
| How should it be exposed? | `Context.ExecutionGuid` |
## Implementation

Every single call through a Polly policy carries an instance of `Context` with it: this will carry the values of the key for each call.  
# Comments?

Comments?  Alternative suggestions?  Extra considerations to bear in mind?  Do we need any other keys?  What do comparable products provide in their stats?

**Comment by SeanFarrow** [comment]
Any idea who is working on this/when they will be starting?

**Comment by reisenberger** [comment]
Hi @SeanFarrow . I'll probably take this one on myself, as passing the context/keys on correctly down the wrap will be closely aligned with the work on `PolicyWrap`.  Are you waiting on me to add `ExecutionKey` to drive lookups from/storing in the cache for #136?  

I can possibly start this (or we can put in a stub) in the coming week.  Shout at any point if you're ready to start and this is holding you up!

**Comment by SeanFarrow** [comment]
Hi,

I am waiting for the ExecutionKey, it’s not critical as yet as I’m starting work on the cache implementations, but may be within a week or so, depending on other things!
Cheers

**Comment by brunolauze** [comment]
Can we describe what are the difference between wording usage and call

...for each usage of a policy.
...for each call through a policy...

**Comment by reisenberger** [comment]
Hi @brunolauze.  Good question.  Say a single `RetryPolicy` is used to wrap outbound calls at three call sites in the code:

[a] Calls from all three call sites would share the same `PolicyKey` (the same `Policy` instance is in use)

[b] Calls from each different call site would get an individual `ExecutionKey` _if the coder specified it when coding the call to_ `Execute/Async(...)`.  Each time that call site places a call, that `ExecutionKey` will apply.

_Aside:_ The `ExecutionKey` would be specified by using an Execute overload taking a Context, passing the executionKey to Context thus: `.Execute/Async(whatever, new Context(executionKey))`

[c] Each time any call site places a call, a random `ExecutionGuid` would be assigned (unique per runtime call).

[d] `PolicyWrapKey`s operate like `PolicyKey`s - one per instance - so the same where the same `PolicyWrap` inytance is used, different where different instances are used.

So:

| Key  | Scope |
| ------------- | ------------- |
| PolicyWrapKey | per PolicyWrap instance |
| PolicyKey | per Policy instance |
| ExecutionKey | per call site in the codebase, _if specified by the coder when coding_ |
| ExecutionGuid | per call at runtime |

Thanks for calling out that this could be clearer: I'll update the original!

**Comment by reisenberger** [comment]
Hey @brunolauze / everyone.  I've updated the [original table of Policy Keys](https://github.com/App-vNext/Polly/issues/139#issue-164897820) with an extra row 'How would it be specified?', to show that.  

_Usages_ would be:

+ All values would travel on the `Context` travelling with each execution
+ Coders can select `onRetry`, `onBreak`, `onTimeout` etc etc delegates taking the `Context`, so if eg these are used for logging, loggers can log as much context as you like.
+ When Polly is extended to emit stats/events, it's intended that all these keys would also feature in the `PolicyEvent`s emitted.  So event aggregators/dashboards could identify trends per policy, policy wrap, policy usage (call site), determine latency of individual executions and aggregate this to give averages, etc.

Any other questions / ideas, fire away!

**Comment by reisenberger** [comment]
Just to state that `PolicyKey`s and their relation to the execution `Context` has been delivered and merged into the [v5.0-alpha branch](https://github.com/App-vNext/Polly/tree/v5.0-alpha) at [this commit](https://github.com/App-vNext/Polly/commit/e1d2979d9117cc1d01d2d2723f31ff524eccd3b7).

@dgodd @SeanFarrow This provides the full TDD-driven approach to keys underpinning execution `Context` for both metrics, and cache keys.

Only _difference_ from the original proposal: it felt redundant (when implemented) for a `PolicyWrap` to have both a `PolicyKey` and a `PolicyWrapKey` property.  

The  `PolicyWrapKey` property on the _execution_ `Context` is essential, and remains the same: it carries as a constant through the execution of any (however-nested) `PolicyWrap`, but a `Context` just takes the `PolicyKey` of the outermost `PolicyWrap` to be the `Context.PolicyWrapKey` property.
